### PR TITLE
AJAX User Logout Part 2

### PR DESCRIPTION
**Pull request by rajadain**
Connects #173 

Revises #201 

I'm creating a new PR because my old private repo was blown away as a part of the transition. For reference, the old PR is #201.

**Testing Notes**
Login as a user. Click name in the header and logout. Verify that the username now has changed to a Login button. Click the login button to make the login modal show up. Refresh the page without logging in, and it should identify you as not logged in and pop up the login modal.
Login as a user. Right click the logout link and open in new tab. Refresh the first tab, and it should identify you as not logged in and pop up the login modal.

**Further Work**
 * This PR has an unstyled log out page. This probably needs some design attention, and input from @mmcfarland and @ctaylo37 for how best to handle. Created #214 for this.
 * The original issue asks for the user to be redirected the home page in case they are at a page which is for logged-in users only. I don't have a clear idea of which pages this would apply to, and once we have that we will have to figure out how to mark those pages as requiring authentication within Backbone / Marionette. Created #215 for this.

### CODE REVIEW

**Comment by rajadain** [comment]
Connects #173

File: src/mmw/apps/user/templates/user/logout.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{% extends 'base.html' %}
+
+{% block content %}
+    <h1>You have been logged out.</h1>
+
+{% endblock content %}
```
Comment (kdeloach) [inline src/mmw/apps/user/templates/user/logout.html]: I would get an opinion from @ctaylo37 to see if we should implement a separate logout page. In other projects we redirect to the login page with `?next_url=` set to whatever page you were just looking at.

File: src/mmw/apps/user/views.py [hunk 2]
```diff
@@ -47,6 +47,18 @@ def ajax_login(request):
...
+@decorators.permission_classes((AllowAny, ))
+def user_logout(request):
+    logout(request)
+
+    if request.is_ajax():
+        response_data = {'guest': True, 'result': 'success'}
+        return Response(data=response_data)
+    else:
+        return render_to_response('user/logout.html')
+
```
Comment (kdeloach) [inline src/mmw/apps/user/views.py]: I like this much better than having to maintain separate urls/views for ajax vs non-ajax requests. We should consider doing the same for "ajaxlogin" in the future.

File: src/mmw/js/src/user/models.js [hunk 3]
```diff
@@ -1,13 +1,54 @@
...
+    // be specified upon usage. They both update the user model, so any event
+    // listeners that subscribe to the `sync` event will be triggered.
+    login: function(attrs) {
+        return this.fetch({
+            type: 'POST',
+            url: 'user/ajaxlogin',
+            data: attrs,
+            beforeSend: function(xhr) {
+                var csrftoken = helpers.Cookie.get('csrftoken');
+                xhr.setRequestHeader('X-CSRFToken', csrftoken);
```
Comment (kdeloach) [inline src/mmw/js/src/user/models.js]: Consider rebasing on top of #213 to prevent merge conflicts.

**Comment by kdeloach** [comment]
+1

**Comment by kdeloach** [comment]
I merged #213 so you should be able to rebase on top of develop now.

**Comment by rajadain** [comment]
Rebased on top of @kdeloach's work on #213 by removing inline CSRF references, now that they have been configured globally.

**Comment by lliss** [comment]
Is it done elsewhere in the app? It seems fine to me.

**Comment by lliss** [comment]
+1. Tested.

**Comment by caseypt** [comment]
I think we should avoid using `window.MMW.*` in our code. The purpose was to have it there for debugging. I don't believe we are doing this else where in the application.

We should consider refactoring our `main.js`, `app.js`, etc to be able to get access to the global objects we need through `require` and avoid have circular imports.

**Comment by mmcfarland** [comment]
I agree about accessing off of `window` for the same reason we avoid other global variables.  Refactoring to support requiring seems like the way to go.

**Comment by rajadain** [comment]
I've added two options:

 1. d96f969 simply requires `router` in `core/views`
 2. 0634bb0 is a more elaborate refactoring, in which we move `router` from `main` to `App`, thus giving everyone who has access to `App` to be able to use the `router`. Additionally, we inject `App` into `HeaderView` at initialization, allowing it to call methods defined in `App` without having to `require` it inline.

Honestly I'm unsure which is the better one. The first one is easy and simple, but is it too simplistic? The second one does a slightly better job of reducing `require` statements, since instead of doing it in two places `main` and `core/views` we now do it in only one place `App`, and reference it or inject it. But does this cause unnecessary coupling between these modules? Is the best way somewhere in between?

**Comment by lliss** [comment]
I prefer the first solution.

File: src/mmw/apps/user/views.py (hunk 2 above)
Comment (lliss) [inline src/mmw/apps/user/views.py]: I agree. This is nice.

**Comment by rajadain** [comment]
Very well, I've squashed the first solution into the main branch and discarded the alternative. Will merge after a final okay.

File: src/mmw/js/src/core/templates/header.ejs [hunk 4]
```diff
@@ -13,7 +13,7 @@
                     </a>
                     <ul class="dropdown-menu">
                         <li><a href="#">Profile</a></li>
-                        <li><a href="accounts/logout">Logout</a></li>
+                        <li><a href="/user/logout" class="user-logout" onclick="return false;">Logout</a></li>
```
Comment (lliss) [inline src/mmw/js/src/core/templates/header.ejs]: I don't think this really bothers anyone else, but I prefer keep the html pure and then call ```preventDefault``` in the handler.

**Comment by lliss** [comment]
+1. Tested.

File: src/mmw/js/src/core/templates/header.ejs (hunk 4 above)
Comment (mmcfarland) [inline src/mmw/js/src/core/templates/header.ejs]: Agreed.  I think event handlers should be implemented in js files, not html/templates, so it's easier to find and read the behavior.

**Comment by rajadain** [comment]
Switched to using `preventDefault` in the handler in 4d287d1. Since that was the only issue and it is now fixed, seems like this is ready to merge!

**Comment by lliss** [comment]
Great. Nice work.

**Pull request merged**